Spigot-Server-Patches/0491-Implement-JellySquid-s-Entity-Collision-optimisation.patch

```diff
@@ -0,0 +1,397 @@
++
++import net.minecraft.server.*;
++import java.util.*;
++
++import java.util.List;
++import java.util.function.Consumer;
++import java.util.stream.Stream;
++import java.util.stream.StreamSupport;
++
++public class LithiumEntityCollisions {
++    /**
++      * [VanillaCopy] CollisionView#getBlockCollisions(Entity, Box)
++      * This is a much, much faster implementation which uses simple collision testing against full-cube block shapes.
++      * Checks against the world border are replaced with our own optimized functions which do not go through the
++      * VoxelShape system.
++     */
++    public static Stream<VoxelShape> getBlockCollisions(ICollisionAccess world, final Entity entity, AxisAlignedBB entityBox) {
++        int minX = MathHelper.floor(entityBox.minX - 1.0E-7D) - 1;
++        int maxX = MathHelper.floor(entityBox.maxX + 1.0E-7D) + 1;
++        int minY = MathHelper.floor(entityBox.minY - 1.0E-7D) - 1;
++        int maxY = MathHelper.floor(entityBox.maxY + 1.0E-7D) + 1;
++        int minZ = MathHelper.floor(entityBox.minZ - 1.0E-7D) - 1;
++        int maxZ = MathHelper.floor(entityBox.maxZ + 1.0E-7D) + 1;
++
++        final VoxelShapeCollision context = entity == null ? VoxelShapeCollision.a() : VoxelShapeCollision.a(entity);
++        final CursorPosition cuboidIt = new CursorPosition(minX, minY, minZ, maxX, maxY, maxZ);
++        final BlockPosition.MutableBlockPosition pos = new BlockPosition.MutableBlockPosition();
++        final VoxelShape entityShape = VoxelShapes.of(entityBox);
++
++        return StreamSupport.stream(new Spliterators.AbstractSpliterator<VoxelShape>(Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.IMMUTABLE) {
++            boolean skipWorldBorderCheck = entity == null;
++
++            public boolean tryAdvance(Consumer<? super VoxelShape> consumer) {
++                if (!this.skipWorldBorderCheck) {
++                    this.skipWorldBorderCheck = true;
++
++                    WorldBorder border = world.getWorldBorder();
++                    boolean isInsideBorder = LithiumEntityCollisions.isBoxFullyWithinWorldBorder(border, entity.getBoundingBox().shrink(1.0E-7D));
++                    boolean isCrossingBorder = LithiumEntityCollisions.isBoxFullyWithinWorldBorder(border, entity.getBoundingBox().grow(1.0E-7D));
++                    if (!isInsideBorder && isCrossingBorder) {
++                        consumer.accept(border.asVoxelShape());
++
++                        return true;
++                    }
++                }
++
++                while (cuboidIt.step()) {
++                    int x = cuboidIt.getX();
++                    int y = cuboidIt.getY();
++                    int z = cuboidIt.getZ();
++
++                    int edgesHit = cuboidIt.getEdgeCoordinatesCount();
++
++                    if (edgesHit == 3) {
++                        continue;
++                    }
++
++                    IBlockAccess chunk = world.c(x >> 4, z >> 4); // c() = getExistingChunk()
```

Comment:
consider https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0454-Optimize-Collision-Chunk-lookup-and-avoid-loading-fa.patch
here

I didn't know enough about this system, but if the chunk is null, we really should be returning a full block collision to stop the movement entirely into the unloaded chunk.

Returning null here would consider it as a non collision, can we fix it to mark it as a collision?